src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticRangeVisitor.cs

```diff
@@ -384,9 +384,25 @@ private void AddSemanticRange(SyntaxNode node, int semanticKind)
             var range = node.GetRange(source);
 
             // LSP spec forbids multi-line tokens, so we need to split this up.
-            // Thankfully all instances of this have multiple component tokens.
             if (range.Start.Line != range.End.Line)
             {
+                var childNodes = node.ChildNodes();
+                if (childNodes.Count == 0)
+                {
+                    var content = node.GetContent();
+                    var lines = content.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
+                    var charPosition = range.Start.Character;
+                    for (var i = 0; i < lines.Length; i++)
+                    {
+                        var startPosition = new Position(range.Start.Line + i, charPosition);
+                        var endPosition = new Position(range.Start.Line + i, charPosition + lines[i].Length);
+                        var lineRange = new Range(startPosition, endPosition);
+                        var semantic = new SemanticRange(semanticKind, lineRange, modifier: 0);
+                        AddRange(semantic);
+                        charPosition = 0;
+                    }
+                }
+
                 // We have to iterate over the individual nodes because this node might consist of multiple lines
                 // ie: "/r/ntext/r/n" would be parsed as one node containing three elements (newline, "text", newline).
                 foreach (var token in node.ChildNodes())
```

Comment:
Hmm, wont this create additional semantic ranges for the whitespace portions of comments?

Reply:
@NTaylorMullen yes. That displays just fine in both VS and VSCode, but it does present some inefficiencies. We could attempt to "meld" tokens like this that are one the same line, but I get nervous about trying to introspect on node contents too much.

Reply:
Oh, I read this as being a general question the first time but you clearly called out comments.

It will create multiple tokens in every multi-line case EXCEPT razor comments. For whatever reason Razor Comments are the only part of the syntax tree that AREN'T tokenized the way you describe.

Reply:
> It will create multiple tokens in every multi-line case EXCEPT razor comments. For whatever reason Razor Comments are the only part of the syntax tree that AREN'T tokenized the way you describe.

holy shit really?? That's awful! @pranavkm @captainsafia this is something we probably will want to change in the parser otherwise we'll be allocating loads.

@ryanbrandenburg I totally get why you're doing it then. If we find users frequently complaining about this from a perf perspective we'll need to be smarter in how we iterate the comment string then 😢, aka iterate character by character looking for newlines etc.

Reply:
Hmmmm. I'm unfamiliar with this problem. Mind opening an issue over on dotnet/aspnetcore with some details?

Reply:
I filed https://github.com/dotnet/aspnetcore/issues/31575. I'm going to do what I have for now and we can revert that section if/when the compiler agrees with me.